File: CarDemo.js
// Defining the Engine interface
interface Engine {
    start();
    stop();
}

// Defining the Vehicle abstract class
class Vehicle {
    constructor(company, model, year) {
        this.company = company;
        this.model = model;
        this.year = year;
    }

    // Abstract method drive
    drive() {}
    
    // Method to display vehicle information
    displayInfo() {
        console.log("Vehicle Information: " + this.company + " " + this.model + " " + this.year);
    }
}

// Defining the Car class which extends Vehicle and implements Engine
class Car extends Vehicle {
    constructor(company, model, year, fuelTank) {
        super(company, model, year);
        this.fuelTank = fuelTank;
        this.isEngineRunning = false;
    }

    // Overridden drive method
    drive() {
        if (this.isEngineRunning) {
            if (this.fuelTank.getCurrentFuelLevel() > 0) {
                console.log("The car can be drive.");
                this.fuelTank.consumeFuel(10); 
            } else {
                console.log("car can't drive");
            }
        } else {
            console.log("Start the car before driving.");
        }
    }

    // Overridden start method from Engine interface
    start() {
        if (!this.isEngineRunning && this.fuelTank.getCurrentFuelLevel() > 0) {
            this.isEngineRunning = true;
            console.log("The car engine is now started.");
        } else {
            console.log("Unable to start the car engine.");
        }
    }

    // Overridden stop method from Engine interface
    stop() {
        if (this.isEngineRunning) {
            this.isEngineRunning = false;
            console.log("The car engine is now stopped.");
        } else {
            console.log("The car engine is already stopped.");
        }
    }
}

// Defining the ElectricCar class which extends Vehicle and implements Engine
class ElectricCar extends Vehicle {
    constructor(company, model, year) {
        super(company, model, year);
        this.batteryLevel = 100;
        this.isEngineRunning = false;
    }

    // Overridden drive method
    drive() {
        if (this.isEngineRunning) {
            if (this.batteryLevel > 0) {
                console.log("The electric car is now driving.");
                this.batteryLevel -= 10; 
            } else {
                console.log("need to charge battery");
            }
        } else {
            console.log("Start the electric car before driving.");
        }
    }

    // Overridden start method from Engine interface
    start() {
        if (!this.isEngineRunning && this.batteryLevel > 0) {
            this.isEngineRunning = true;
            console.log("The electric car engine is now started.");
        } else {
            console.log("Unable to start the electric car engine. Check battery level or engine status.");
        }
    }

    // Overridden stop method from Engine interface
    stop() {
        if (this.isEngineRunning) {
            this.isEngineRunning = false;
            console.log("The electric car engine is now stopped.");
        } else {
            console.log("The electric car engine is already stopped.");
        }
    }
}

// Defining the FuelTank interface
interface FuelTank {
    getCurrentFuelLevel();
    consumeFuel(amount);
}

// Defining the FuelTank1 class which implements FuelTank interface
class FuelTank1 {
    constructor(capacity) {
        this.capacity = capacity;
        this.currentFuelLevel = capacity; 
    }

    // Method to get current fuel level
    getCurrentFuelLevel() {
        return this.currentFuelLevel;
    }

    // Method to consume fuel
    consumeFuel(amount) {
        if (this.currentFuelLevel >= amount) {
            this.currentFuelLevel -= amount;
            console.log(amount + " liters fuel used. Current fuel level: " + this.currentFuelLevel);
        } else {
            console.log("Not enough fuel to consume.");
        }
    }
}

// Main function
function carDemo() {
    const carFuelTank = new FuelTank1(50);
    const myCar = new Car("Toyota", "Fortuner", 2022, carFuelTank);
    
    const myElectricCar = new ElectricCar("Maruti", "Baleno", 2022);
    
    myCar.displayInfo();
    myCar.start();
    myCar.drive();
    myCar.displayInfo();
    myCar.stop();
    myCar.displayInfo();
    myCar.drive();
    myCar.displayInfo();
    
    myElectricCar.displayInfo();
    myElectricCar.start();
    myElectricCar.drive();
    myElectricCar.displayInfo();
    myElectricCar.stop();
    myElectricCar.displayInfo();
}

// Calling the main function
carDemo();

File: CarDemo.java
import java.lang.*;
import java.util.Scanner;

interface Engine {
    void start();

    void stop();
}


abstract class Vehicle {
    private String company;
    private String model;
    private int year;

    public Vehicle(String company, String model, int year) {
        this.company = company;
        this.model = model;
        this.year = year;
    }

    
    public abstract void drive();   //abstract method

    
    public void displayInfo() {
        System.out.println("Vehicle Information: " + company + " " + model + " " + year);
    }
}

class Car extends Vehicle implements Engine {    //inheritance and interface implement
    private FuelTank fuelTank;
    private boolean isEngineRunning;

    public Car(String company, String model, int year, FuelTank fuelTank) {
        super(company, model, year);
        this.fuelTank = fuelTank;
        this.isEngineRunning = false;
    }

    
    public void drive() {      //override abstract method
        if (isEngineRunning) {
            if (fuelTank.getCurrentFuelLevel() > 0) {
                System.out.println("The car can be drive.");
                fuelTank.consumeFuel(10); 
            } 
            else {
                System.out.println("car can't drive");
            }
        } else {
            System.out.println("Start the car before driving.");
        }
    }

    public void start() {       //override method of interface
        if (!isEngineRunning && fuelTank.getCurrentFuelLevel() > 0) {
            isEngineRunning = true;
            System.out.println("The car engine is now started.");
        } else {
            System.out.println("Unable to start the car engine.");
        }
    }

    
    public void stop() {     //override method of interface
        if (isEngineRunning) {
            isEngineRunning = false;
            System.out.println("The car engine is now stopped.");
        } else {
            System.out.println("The car engine is already stopped.");
        }
    }
}


class ElectricCar extends Vehicle implements Engine {
    private int batteryLevel;
    private boolean isEngineRunning;

    public ElectricCar(String company, String model, int year) {
        super(company, model, year);
        this.batteryLevel = 100; 
        this.isEngineRunning = false;
    }

    
    public void drive() {         //override
        if (isEngineRunning) {
            if (batteryLevel > 0) {
                System.out.println("The electric car is now driving.");
                batteryLevel -= 10; 
            } else {
                System.out.println("need to charge battery");
            }
        } else {
            System.out.println("Start the electric car before driving.");
        }
    }

    
    public void start() {   //override
        if (!isEngineRunning && batteryLevel > 0) {
            isEngineRunning = true;
            System.out.println("The electric car engine is now started.");
        } else {
            System.out.println("Unable to start the electric car engine. Check battery level or engine status.");
        }
    }

    
    public void stop() {   //override
        if (isEngineRunning) {
            isEngineRunning = false;
            System.out.println("The electric car engine is now stopped.");
        } else {
            System.out.println("The electric car engine is already stopped.");
        }
    }
}


interface FuelTank {            //interface for fuel
    int getCurrentFuelLevel();
    void consumeFuel(int amount);
   // void refuel(int amount);
}

class FuelTank1 implements FuelTank {
    private int capacity;
    private int currentFuelLevel;

    public FuelTank1(int capacity) {
        this.capacity = capacity;
        this.currentFuelLevel = capacity; 
    }

    
    public int getCurrentFuelLevel() {   //override
        return currentFuelLevel;
    }

    
    public void consumeFuel(int amount) {   //override
        if (currentFuelLevel >= amount) {
            currentFuelLevel -= amount;
            System.out.println(amount + " liters fuel used. Current fuel level: " + currentFuelLevel);
        } else {
            System.out.println("Not enough fuel to consume.");
        }
    }

    
    /*public void refuel(int amount) {      //override
        if (currentFuelLevel + amount <= capacity) {
            currentFuelLevel += amount;
        } else {
            currentFuelLevel = capacity;
        }
    }*/
}

public class CarDemo {
    public static void main(String[] args) {
        
        FuelTank1 carFuelTank = new FuelTank1(50);
        Car myCar = new Car("Toyota", "Fortuner", 2022, carFuelTank);

    
        ElectricCar myElectricCar = new ElectricCar("Maruti", "Baleno", 2022);

        myCar.displayInfo();
        myCar.start();
        myCar.drive();
        myCar.displayInfo();
        myCar.stop();
        myCar.displayInfo();
       // myCar.refuel(20);
        myCar.drive();
        myCar.displayInfo();

        myElectricCar.displayInfo();
        myElectricCar.start();
        myElectricCar.drive();
        myElectricCar.displayInfo();
        myElectricCar.stop();
        myElectricCar.displayInfo();
    }
}